Asterix/optics/propagation_functions.py

```diff
@@ -610,30 +615,81 @@ def prop_fpm_regional_sampling(pup,
-                         f"spatial frequency content in the outer sampling layer.")
+    nbres = np.array(nbres)
 
-    dim_pup = pup.shape[0]
+    dim_overpad_pupil = pup.shape[0]
     dim_fpm = fpm.shape[0]
-    pup0 = np.array(pup, copy=True, dtype='complex128')
+
+    if real_dim_input is None:
+        real_dim_input = pup.shape[0]
+    samplings = real_dim_input / nbres
+
+    if not np.all(np.diff(nbres) >= 0):  # check if it is sorted by checking 2 by 2 diff is always positive
+        raise ValueError(f"'nbres' parameter need to be sorted from the highest to lowest nbrs of elements."
+                         f"Currently 'nbres' = {nbres}")
+
+    if np.min(samplings) < 2:
+        raise ValueError(f"The outer sampling in prop_fpm_regional_sampling is hardcoded 2, otherwise we cut off"
+                         f"the high-spatial frequencies and the simulation turns out bad. We need the samplings"
+                         f"defined by the 'nbres' parameter (real_dim_input/nbres) to be always >= 2. Currently, with"
+                         f"real_dim_input = {real_dim_input}, samplings are {samplings}")
+
+    if np.min(samplings) == 2:
+        # If the outer sampling defined by nbrs is already 2, we can remove it and gain some time
+        # because the last sampling is harcoded at 2
+        nbres = np.delete(nbres, -1)
+        samplings = np.delete(samplings, -1)
+
+    if max(shift) >= nbres[0] / 2:
+        raise ValueError(f"shift {shift} is larger than the minimum nbrs {nbres[0]} of element of resolution : the "
+                         f"tip/tilt is out of the array! Increase min(nbres) or decrease shift.")
+
+    # can be used to check:
+    # print(f"With real_dim_input = {real_dim_input} and nbrs = {nbres}, Samplings: ", samplings)
 
     # Add tip-tilt to in pupil wavefront to simulate FPM offsets
-    phase_ramp = shift_phase_ramp(dim_pup, shift[0], shift[1])
-    inverted_phase_ramp = shift_phase_ramp(dim_pup, -shift[0], -shift[1])
+    # phase_ramp = shift_phase_ramp(dim_pup, shift[0], shift[1])
+    # inverted_phase_ramp = shift_phase_ramp(dim_pup, -shift[0], -shift[1])
+
+    # non non non ! shift_phase_ramp function parameters shift_x and shift_y are in pixels!
+    # Doing tip/tilt like that, you introduce a fix pixel shift. But because the resolution change,
+    # the tip_tilt change also for each samplings.
+    # For example: if samplings = 200, 20, 2 (pix per lambda /D), a shift = (1,0) (1 pix shift) will correspond to a
+    # tip tilt of 0.005 l/D in the first array, 0.05 l/D in the second and 0.5 l/D in the last
+    # I see that you introduced the 'shift' parameter in the docstring in element of resolution, which is
+    # a good idea to avoid that. See below the correct way (I think) to do that, we need to multiply each time by
+    # sampling to have the right shift in pixel. I also use the parameter X_offset_output and X_offset_input which
+    # are exactly doing that
```

Comment:
Could we delete this? @rgalicher and I convinced ourselves that this was fine. I am happy to revisit together, but maybe it's not the time to add this comment as if it was a done deal. You could instead write something like: "Shifting needs to be revisited due to confusion about shift units." or whatever you like.

Or, since you changed the way the shifting is done, just delete all this commented code altogether.

Reply:
done